drivers/network/netio/netio.c

```diff
@@ -0,0 +1,904 @@
+                            }
+                            if (InputSize < sizeof(WSK_EVENT_CALLBACK_CONTROL))
+                            {
+                                DbgPrint("WskControlSocket: InputBuffer too small for this operation\n");
+                                status = STATUS_INVALID_PARAMETER;
+                                break;
+                            }
+                            WSK_EVENT_CALLBACK_CONTROL *c = (WSK_EVENT_CALLBACK_CONTROL *) InputBuffer;
+
+                            s->CallbackMask = c->EventMask;
+                            /* TODO: and start listening here? */
+
+                            status = STATUS_SUCCESS;
+                            break;
+
+                        default:
+                            DbgPrint("WskControlSocket: ControlCode %d Not implemented\n", ControlCode);
+                    }
+                    break;
+                default:
+                    DbgPrint("WskControlSocket: Level %d Not implemented\n", Level);
+            }
+            break;
+
+        case WskGetOption:
+        case WskIoctl:
+        default:
+            DbgPrint("WskControlSocket: Option %d Not implemented\n", RequestType);
+    }
+    if (Irp != NULL)
+    {
+        Irp->IoStatus.Status = status;
+    }
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskCloseSocket(_In_ PWSK_SOCKET Socket, _Inout_ PIRP Irp)
+{
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+
+    SocketPut(s);
+    /* TODO: DummyCallDriver and IoComplete */
+    return STATUS_SUCCESS;
+}
+
+static struct _TRANSPORT_ADDRESS *
+TdiTransportAddressFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    struct _TRANSPORT_ADDRESS *ta;
+
+    ta = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ta) + sizeof(struct sockaddr), 'ADDR');
+    if (ta == NULL)
+    {
+        DbgPrint("TdiTransportAddressFromSocketAddress: Out of memory\n");
+        return NULL;
+    }
+
+    ta->TAAddressCount = 1;
+    ta->Address[0].AddressLength = sizeof(SocketAddress->sa_data);
+    ta->Address[0].AddressType = SocketAddress->sa_family;
+    memcpy(&ta->Address[0].Address[0], &SocketAddress->sa_data, ta->Address[0].AddressLength);
+
+    return ta;
+}
+
+static PTDI_CONNECTION_INFORMATION
+TdiConnectionInfoFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    PTRANSPORT_ADDRESS TargetAddress;
+    PTDI_CONNECTION_INFORMATION ConnectionInformation = NULL;
+    NTSTATUS status;
+
+    TargetAddress = TdiTransportAddressFromSocketAddress(SocketAddress);
+    if (TargetAddress == NULL)
+        return NULL;
+
+    status = TdiBuildConnectionInfo(&ConnectionInformation, TargetAddress);
+    ExFreePoolWithTag(TargetAddress, 'ADDR');
+
+    if (!NT_SUCCESS(status))
+    {
+        if (ConnectionInformation != NULL)
+        {
+            ExFreePoolWithTag(ConnectionInformation, TAG_AFD_TDI_CONNECTION_INFORMATION);
+        }
+        return NULL;
+    }
+    return ConnectionInformation;
+}
+
+
```

Comment:
```suggestion
```

Reply:
Applied, thanks :)

Reply:
Resolved all and everything. Inviting others now to have a closer look at it from a functional pov. I am back to neutrality. Thank you for going the extra-length.

Reply:
You're welcome, thanks for the review :)